src/js_parser.zig

```diff
@@ -23102,677 +23059,355 @@ fn NewParser_(
+                    // useState first argument is initial state.
+                    .useState => 0,
+                    // useReducer second argument is initial state.
+                    .useReducer => 1,
+                    else => break :hash_arg,
+                };
+                if (hook_call.args.len <= arg_index) break :hash_arg;
+                const arg = hook_call.args.slice()[arg_index];
+                arg.data.writeToHasher(&ctx.hasher, p.symbols.items);
+            } else switch (hook_call.target.data) {
+                inline .e_identifier,
+                .e_import_identifier,
+                .e_commonjs_export_identifier,
+                => |id| {
+                    const gop = ctx.user_hooks.getOrPut(p.allocator, id.ref) catch bun.outOfMemory();
+                    if (!gop.found_existing) {
+                        gop.value_ptr.* = Expr.initIdentifier(id.ref, logger.Loc.Empty);
                     }
-                }
+                },
+                else => {},
             }
 
-            switch (commonjs_wrapper_expr) {
-                .bun_dev => |commonjs_wrapper| {
-                    var require_function_args = allocator.alloc(Arg, 2) catch unreachable;
-                    var final_part_stmts_count: usize = 0;
+            ctx.hasher.update("\x00");
+        }
 
-                    var imports_count: u32 = 0;
-                    // We have to also move export from, since we will preserve those
-                    var exports_from_count: u32 = 0;
+        pub fn handleReactRefreshPostVisitFunctionBody(p: *P, stmts: *ListManaged(Stmt), hook: *ReactRefresh.HookContext) void {
+            bun.assert(p.options.features.react_fast_refresh);
 
-                    // Two passes. First pass just counts.
-                    for (parts) |part| {
-                        for (part.stmts) |stmt| {
-                            imports_count += switch (stmt.data) {
-                                .s_import => @as(u32, 1),
-                                else => @as(u32, 0),
-                            };
-
-                            exports_from_count += switch (stmt.data) {
-                                .s_export_star, .s_export_from => @as(u32, 1),
-                                else => @as(u32, 0),
-                            };
+            // we need to prepend `_s();` as a statement.
+            if (stmts.items.len == stmts.capacity) {
+                // would have to re-alloc the array in general
+                const new_stmts = p.allocator.alloc(Stmt, stmts.items.len + 1) catch bun.outOfMemory();
+                @memcpy(new_stmts[1..], stmts.items);
+                stmts.deinit();
+                stmts.* = ListManaged(Stmt).fromOwnedSlice(p.allocator, new_stmts);
+            } else {
+                _ = stmts.addOneAssumeCapacity();
+                bun.copy(Stmt, stmts.items[1..], stmts.items[0 .. stmts.items.len - 1]);
+            }
```

Comment:
```
stmts.ensureUnusedCapacity(1);
bun.copy(...);
```
?

Reply:
the slice already has enough capacity. addOne increases the length.

annotating this with more comments with my rationale

```zig
// We need to prepend `_s();` as a statement.
if (stmts.items.len == stmts.capacity) {
    // If the ArrayList does not have enough capacity, it is
    // re-allocated entirely to fit. Only one slot of new capacity
    // is used since we know this statement list is not going to be
    // appended to afterwards; This function is a post-visit handler.
    const new_stmts = p.allocator.alloc(Stmt, stmts.items.len + 1) catch bun.outOfMemory();
    @memcpy(new_stmts[1..], stmts.items);
    stmts.deinit();
    stmts.* = ListManaged(Stmt).fromOwnedSlice(p.allocator, new_stmts);
} else {
    // The array has enough capacity, so there is no possibility of
    // allocation failure. We just move all of the statements over
    // by one, and increase the length using `addOneAssumeCapacity`
    _ = stmts.addOneAssumeCapacity();
    bun.copy(Stmt, stmts.items[1..], stmts.items[0 .. stmts.items.len - 1]);
}
```